### PR TITLE
New app for playing golf, Golf GPS

### PR DESCRIPTION
**Pull request by jeonlab**
A golf GPS app. 
Add as many golf course coordinate data as you want and play, keep score, show distances to the hole and from the last shot. Store score and you can browse them later.

### CODE REVIEW

**Comment by bobrippling** [comment]
@jeonlab I'll mark this as a draft for the time being while changes are being pushed - let me know when you're ready though

**Comment by jeonlab** [comment]
@bobrippling, I thought I was ready, but found a few things to update and correct. Thank you for marking this PR as a draft. I will let you know when I’m done. Thanks.

**Comment by bobrippling** [comment]
Sounds good!

**Comment by jeonlab** [comment]
A failed check has been cleared.

**Comment by jeonlab** [comment]
@bobrippling, I just finished the last update. It's ready for review.
Thanks.

**Comment by bobrippling** [comment]
Thanks - shall check it out soon :)

File: apps/golf-gps/golf-gps.js [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+  if (scoreFiles.length === 0) {
+    E.showPrompt("No score file found.\n\nYou need to play at least a game.", {
+      title: `Error`,
+      buttons: {
+        "END": 1
+      }
+    }).then(choice => {
+      if (choice === 1) load();
+    });
+  }
```
Comment (bobrippling) [inline apps/golf-gps/golf-gps.js:119]: Do you need to `return` after issuing this `E.showPrompt()` ? Otherwise I expect the below browse-files prompt will be shown

File: apps/golf-gps/golf-gps.js [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+    E.showPrompt(l, {
+      buttons: {
+        "<<": 1,
+        ">>": 2,
+        "End": 3
+      }
+    }).then(choice => {
+      if (choice === 1) fileIndx = (fileIndx - 1 + scoreFiles.length) % scoreFiles.length;
+      else if (choice === 2) fileIndx = (fileIndx + 1) % scoreFiles.length;
+      else if (choice === 3) load();
```
Comment (bobrippling) [inline apps/golf-gps/golf-gps.js:135]: I'm happy if you prefer this - if you do like though, you can use a switch:
```javascript
switch(choice) {
  case 1:
    fileIndx = (fileIndx - 1 + scoreFiles.length) % scoreFiles.length;
    break;
  case 2:
    fileIndx = (fileIndx + 1) % scoreFiles.length;
    break;
  case 3:
    load();
    break;
}
```

File: apps/golf-gps/golf-gps.js [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+  if (lastFix.hdop < 2) g.fillRect(100, H / 3, 136, H / 3 + 5);
+  if (lastFix.hdop < 1) g.fillRect(140, H / 3, W, H / 3 + 5);
+}
+
+function finishGame() {
+  let date = new Date();
+  let saveFileContents = ' ';
+  let parTotal = 0;
+
+  Bangle.setGPSPower(0);
```
Comment (bobrippling) [inline apps/golf-gps/golf-gps.js:244]: This probably isn't turning off gps, since we don't specify the name with which we started GPS:

```suggestion
  Bangle.setGPSPower(0, "golf-gps");
```

File: apps/golf-gps/golf-gps.js [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+  const saveFilename = `Scorecard-${date.getFullYear()}${zeroPad(date.getMonth() + 1, 2)}${zeroPad(date.getDate(), 2)}`;
+  const saveFile = require("Storage").open(saveFilename, "w");
+  for (let i = 1; i < 19; i++) {
+    saveFileContents += String(score[i] - par[i]).padStart(2, ' ');
+    saveFileContents += (i == 18 ? ' ' : (i % 6 == 0 ? ' \n ' : ''));
```
Comment (bobrippling) [inline apps/golf-gps/golf-gps.js:249]: If you like, you can avoid the memory overheads of building up a large string and write each part at a time, [using `saveFile.write(...)`](https://www.espruino.com/ReferenceBANGLEJS2#l_StorageFile_write). Although I appreciate the entire contents are displayed in the prompt below

File: apps/golf-gps/playScreen.png
Comment (bobrippling) [inline apps/golf-gps/playScreen.png:1]: This is a really nice UI! And great details in this image too. Out of interest, how did you generate it?

File: apps/messagegui/app.js [hunk 5]
```diff
@@ -505,7 +505,7 @@ function checkMessages(options) {
         // draw the body
         g.drawString(l.join("\n"), x+10,r.y+20+pady);
       }
-      if (!longBody && msg.src) g.setFontAlign(1,1).setFont("6x8").drawString(msg.src, r.x+r.w-2, r.y+r.h-2);
+      if (!longBody && msg.src) g.setFontAlign(-1,1).setFont("6x8").drawString(msg.src, r.x+r.w-2, r.y+r.h-2);
```
Comment (bobrippling) [inline apps/messagegui/app.js:508]: Let's pull this out to another PR - if you could include @thyttan, @gfwilliams, @paul-arg and @rigrig, they're the main contributors so we can run it by them (it will also need a `metadata.json` and changelog bump)

(let me know if you need a hand)

**Review by bobrippling — COMMENTED** [review]
Looks great - just a heads up on the other app changes (gpspoilog, gpssetup and jclock) - usually we do a PR per app, but since one of them is yours and the other two are minor I think it's ok to skip this time

File: apps/golf-gps/golf-gps.js (hunk 1 above)
Comment (jeonlab) [reply apps/golf-gps/golf-gps.js:119]: Well, there is only one button to 'END', so it will go back to the clock when you tap on it. I have tested this case and it worked fine. Any thoughts?

File: apps/golf-gps/golf-gps.js (hunk 2 above)
Comment (jeonlab) [reply apps/golf-gps/golf-gps.js:135]: It works either way, but I like your suggested 'switch' better. Thanks!

File: apps/golf-gps/golf-gps.js (hunk 4 above)
Comment (jeonlab) [reply apps/golf-gps/golf-gps.js:249]: I'm afraid I understand about what you said here. hmmmm... I will take another look later. Thanks.

File: apps/golf-gps/playScreen.png
Comment (jeonlab) [reply apps/golf-gps/playScreen.png:1]: I just take a few screenshots and then put them in a PowerPoint page and added texts. Nothing :-). 
Thanks.

File: apps/messagegui/app.js (hunk 5 above)
Comment (gfwilliams) [reply apps/messagegui/app.js:508]: Yes - this one looks wrong to me anyway - if it's -1 that's left-aligned which means the start coordinate should be at the left of the box like `r.x+2`, not the right as in `r.x+r.w-2`

File: apps/golf-gps/README.md [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+- How to change holes and add/subtract shots
+  - Swipe left/right to change the hole to next/previous (you can move to any hole to update your shots in case you entered wrong number of shots by mistake)
+  - Swipe up/down to add/subtract the number of shots. This will update the total number of shots as well as current shots.
+- After the game
+  - Press the button to either finish the game or go back to the play screen (if you pressed the button by accident).
+  - If you choose to finish the game, it will show the summary of the score with 3x6 matix, shots - par. For example, -1 is for birdie and 0 is for par, and +2 is for double bogey. It also shows the total shots and total par as well.
+
+## Course data
+Before you run Golf GPS, you need to save your favourite golf course data into the storage area.
+Download `storage-write.js` and open it from Espruino Web IDE (https://www.espruino.com/ide/) and enter your course data and run in RAM mode.
```
[the remark this answers is not in the document]
Comment (bobrippling) [reply apps/golf-gps/README.md:27]: No worries - if you want to try a draft and talk things through, @ me on a draft PR :)

File: apps/golf-gps/golf-gps.js (hunk 1 above)
Comment (bobrippling) [reply apps/golf-gps/golf-gps.js:119]: I might have mangled my question a little - what I'm wondering is:
```diff
 function browseScore() {
   const scoreFiles = require("Storage").list(/^Scorecard-/);
   if (scoreFiles.length === 0) {
     E.showPrompt("No score file found.\n\nYou need to play at least a game.", {
       ...
     }).then(choice => {
       if (choice === 1) load();
     });
 
+    // we get to here after having shown the prompt above
   }
   let fileIndx = scoreFiles.length - 1;
 
   function browseFiles() {
     const browsefile = require("Storage").open(scoreFiles[fileIndx].substring(0, 18), "r");
     const l = browsefile.read(80);
 
     E.showPrompt(l, {
       buttons: {
         "<<": 1,
         ">>": 2,
         "End": 3
       }
     }).then(choice => {
       if (choice === 1) fileIndx = (fileIndx - 1 + scoreFiles.length) % scoreFiles.length;
       else if (choice === 2) fileIndx = (fileIndx + 1) % scoreFiles.length;
       else if (choice === 3) load();
       browseFiles();
     });
   }
 
+  // and so we end up down here, and we'll call browseFiles(),
+  // which will show another prompt - I'd expect this will overwrite the first prompt
   browseFiles();
 }
```

File: apps/golf-gps/golf-gps.js (hunk 2 above)
Comment (bobrippling) [reply apps/golf-gps/golf-gps.js:135]: My pleasure!

File: apps/golf-gps/golf-gps.js (hunk 4 above)
Comment (bobrippling) [reply apps/golf-gps/golf-gps.js:249]: No that's not a problem, it's not important :)

File: apps/golf-gps/playScreen.png
Comment (bobrippling) [reply apps/golf-gps/playScreen.png:1]: Cool - thanks!

File: apps/golf-gps/golf-gps.js (hunk 1 above)
Comment (jeonlab) [reply apps/golf-gps/golf-gps.js:119]: Ah... I see what you meant. Yes, you're right. It should look like this.

```
   }
   let fileIndx = scoreFiles.length - 1;
   return 0;
   function browseFiles() {
```
Thanks.

Comment (bobrippling) [reply apps/golf-gps/golf-gps.js:119]: I'm afraid not - the prompt is asynchronous so the callbacks will run quite a while after the function's finished :)

Comment (bobrippling) [reply apps/golf-gps/golf-gps.js:119]: Interesting that it didn't overwrite - I should have some time to take a look next week if you don't get to the bottom of it in the meanwhile, or if it's not an issue we can close this

(there's just [the final comment](https://github.com/espruino/BangleApps/pull/3467#discussion_r1664728301) below to deal with)

Comment (jeonlab) [reply apps/golf-gps/golf-gps.js:119]: Oh, I see. I will see if I can find some time to take a look today.
Thanks.

Comment (jeonlab) [reply apps/golf-gps/golf-gps.js:122]: I have separated the functions, browseScore() and browseFiles() and changed the variables fileIndx and scoreFiles to global variable. I'm not sure whether this is good solution or not, but with this change, the prompts still works fine and don't see any error in the console.  @bobrippling, please review when you have time. Thanks.

File: apps/messagegui/app.js (hunk 5 above)
Comment (bobrippling) [reply apps/messagegui/app.js:508]: Makes sense - I'll remove this in the merge and we can always add it back in later

File: apps/golf-gps/golf-gps.js (hunk 1 above)
Comment (bobrippling) [reply apps/golf-gps/golf-gps.js:122]: Sounds good, glad to hear the prompt is working still and thanks for your changes!

**Comment by bobrippling** [comment]
Thanks for the app!

**Comment by jeonlab** [comment]
> Thanks for the app!

Thank you for your review!!

**Comment by jeonlab** [comment]
@bobrippling , I just found my mistakes. For the gospoilog and gpssetup apps, I had added Bangle.js 2 just for my test purposes in my branch. I shouldn’t have committed those changes in two metadata.json for those apps. Should I correct them in my branch and do a new PR? Sorry about the mess. Please let me know how I can fix this. Thanks.

**Comment by bobrippling** [comment]
No worries - is that definitely the case they won't work? If so then yes, a new PR is ok :)

**Comment by jeonlab** [comment]
Well, `gpspoilog` works for Bangle.js 2 as well, but `gpssetup` won't work because it was only for the GPS chip in the Bangle.js 1. About two months ago, I was trying to see if I could modify the `gpssetup` for the GPS chip in the Bangle.js 2, but didn't have enough time to dig into it. And then I forgot to remove it from the commit. Actually, I still don't really know how to remove the commit from the web version of github. Back then, I thought using the App Loader was the only way to install and test apps on the watch. That's why I committed those changes in my branch although those apps are not owned by me.
If there is a way to revert the pushes for those two apps, please do so. The authors will be upset when they find I changed their apps without consulting with them. Especially for the `gpssetup`, my change was wrong. So, I will put a new PR for the `gpssetup` right now. Please advise what to do with the `gpspoilog` which is fine to leave with my change (adding Bangle.js 2 in the metadata.json). Thanks.

**Comment by bobrippling** [comment]
Great - good to have the `gpspoilog` change in, in that case :) 

(for anyone following/for reference, merged in #3499)
